pyfarm/agent/http/api/software.py

```diff
@@ -0,0 +1,204 @@
+                            "version": version})
+                except Exception as error:
+                    delay = http_retry_delay()
+                    logger.error(
+                        "Failed to post availability of software %s, "
+                        "version %s to master: %r. Will retry in %s "
+                        "seconds.",
+                        software, version, error, delay)
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                else:
+                    data = yield treq.json_content(response)
+
+                    if response.code == OK:
+                        logger.info("Posted availability of software %s, "
+                                    "version %s to master.",
+                                    software, version)
+                        break
+
+                    elif response.code >= INTERNAL_SERVER_ERROR:
+                        delay = http_retry_delay()
+                        logger.warning(
+                            "Could not post availability of software %s, "
+                            "version %s. The master responded with "
+                            "INTERNAL_SERVER_ERROR.  Retrying in %s "
+                            "seconds.", software, version, delay)
+
+                        deferred = Deferred()
+                        reactor.callLater(delay, deferred.callback, None)
+                        yield deferred
+
+                    else:
+                        logger.error(
+                            "Failed to post availability of software %s, "
+                            "version %s: "
+                            "Unexpected status from server %s. Data: %s",
+                            software, version, response.code, data)
+                        break
+
+        @inlineCallbacks
+        def mark_software_not_available(software, version):
+            url = ("{master_api}/agents/{agent}/software/{software}/"
+                    "versions/{version}").format(
+                        master_api=config.get("master_api"),
+                        agent=config.get("agent_id"),
+                        software=software,
+                        version=version)
+            while True:
+                try:
+                    response = yield delete_direct(url)
+                except Exception as error:
+                    delay = http_retry_delay()
+                    logger.error(
+                        "Failed to remove software %s, version %s from this "
+                        "agent on master: %r. Will retry in %s seconds.",
+                        software, version, error, delay)
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                else:
+                    data = yield treq.content(response)
```

Comment:
Alright makes sense.  That should also avoid errors decoding to json too if there happen to be any.